File: src/services/recovery.js
import fs from 'node:fs';
import path from 'node:path';
import { v4 as uuidv4 } from 'uuid';
import logger from '../logger.js';
import { TRANSCRIPT_DIR, SUMMARY_DIR } from '../config.js';
import { transcribeWithRetry } from './transcription.js';
import { summarizeTranscript } from './summarization.js';
import { convertToWordDoc } from './document.js';
import {
  getSessionsNeedingRecovery,
  getUntranscribedEntries,
  getTranscribedEntries,
  isSessionFullyTranscribed,
  buildTranscriptFromManifest,
  endSession,
  markSessionSummarized,
  getSessionStats,
  cleanupOldSessions,
  markStaleSessions
} from './manifest.js';

// Recovery state
let isRecovering = false;
let recoveryQueue = [];

/**
 * Run recovery for all sessions with untranscribed audio
 * @param {boolean} autoSummarize - Whether to auto-summarize completed sessions
 * @returns {Promise<{recovered: number, failed: number, summarized: number}>}
 */
export async function runRecovery(autoSummarize = true) {
  if (isRecovering) {
    console.log('⚠️ Recovery already in progress');
    return { recovered: 0, failed: 0, summarized: 0 };
  }

  isRecovering = true;
  const recoveryEventId = uuidv4();

  console.log('🔄 Starting audio recovery process...');

  logger.info("Starting audio recovery", {
    extra: {
      footprint: null,
      batch_uuid: null,
      user_id: null,
      event_id: recoveryEventId,
      action: "audio_recovery",
      event: "start"
    }
  });

  let recovered = 0;
  let failed = 0;
  let summarized = 0;

  try {
    // First, clean up old completed sessions
    cleanupOldSessions(7);

    // Get sessions needing recovery
    const sessions = getSessionsNeedingRecovery();

    if (sessions.length === 0) {
      console.log('✅ No sessions need recovery');
      isRecovering = false;
      return { recovered: 0, failed: 0, summarized: 0 };
    }

    console.log(`📋 Found ${sessions.length} session(s) needing recovery`);

    for (const session of sessions) {
      console.log(`\n📁 Processing session ${session.sessionId}...`);

      const stats = getSessionStats(session.sessionId);
      if (stats) {
        console.log(`   Total: ${stats.total}, Pending: ${stats.pending}, Failed: ${stats.failed}, Completed: ${stats.completed}`);
      }

      // Get untranscribed entries for this session
      const untranscribed = getUntranscribedEntries(session.sessionId);

      if (untranscribed.length > 0) {
        console.log(`   🎤 ${untranscribed.length} audio file(s) need transcription`);

        for (const entry of untranscribed) {
          console.log(`   📝 Transcribing ${path.basename(entry.audioPath)}...`);

          const result = await transcribeWithRetry(
            entry.audioPath,
            entry.sessionId,
            entry.userId,
            entry.id,
            3 // Max retries
          );

          if (result.success) {
            recovered++;
            console.log(`   ✅ Transcribed: "${result.text.substring(0, 50)}..."`);
          } else {
            failed++;
            console.log(`   ❌ Failed: ${result.error}`);

            // Stop recovery if we hit quota limits
            if (result.isQuotaError) {
              console.log('   💰 Stopping recovery due to quota limits');
              break;
            }
          }

          // Small delay between transcriptions to avoid rate limiting
          await new Promise(resolve => setTimeout(resolve, 500));
        }
      }

      // Check if session is now fully transcribed
      if (autoSummarize && isSessionFullyTranscribed(session.sessionId)) {
        console.log(`   📄 Session fully transcribed, generating summary...`);

        try {
          const summaryResult = await generateSessionSummary(session.sessionId);
          if (summaryResult.success) {
            summarized++;
            console.log(`   ✅ Summary generated: ${summaryResult.summaryPath}`);
          }
        } catch (err) {
          console.error(`   ❌ Failed to generate summary: ${err.message}`);
        }
      }
    }

    console.log(`\n✅ Recovery complete: ${recovered} transcribed, ${failed} failed, ${summarized} summarized`);

    logger.info("Audio recovery completed", {
      extra: {
        footprint: null,
        batch_uuid: null,
        user_id: null,
        event_id: recoveryEventId,
        action: "audio_recovery",
        event: "complete",
        recovered,
        failed,
        summarized
      }
    });

  } catch (error) {
    logger.error("Audio recovery failed", {
      extra: {
        footprint: null,
        batch_uuid: null,
        user_id: null,
        event_id: recoveryEventId,
        action: "audio_recovery",
        event: "error",
        error_message: error.message
      }
    });
    console.error('❌ Recovery error:', error.message);
  } finally {
    isRecovering = false;
  }

  return { recovered, failed, summarized };
}

/**
 * Generate summary for a completed session
 * @param {string} sessionId - Session to summarize
 * @returns {Promise<{success: boolean, summaryPath: string|null}>}
 */
async function generateSessionSummary(sessionId) {
  // Build transcript from manifest entries
  const transcript = buildTranscriptFromManifest(sessionId);

  if (!transcript || transcript.trim().length === 0) {
    return { success: false, summaryPath: null, error: 'No transcript content' };
  }

  // Write transcript to file
  const [guildId, channelId] = sessionId.split(':');
  const ts = new Date().toISOString().replace(/:/g, '-').split('.')[0];
  const transcriptPath = path.join(TRANSCRIPT_DIR, `${guildId}-${channelId}-${ts}-recovered.log`);

  fs.writeFileSync(transcriptPath, transcript);
  endSession(sessionId, transcriptPath);

  // Create a mock sessionLogs map for the summarization function
  const sessionLogs = new Map();
  sessionLogs.set(sessionId, transcriptPath);

  try {
    // Generate summary
    const summary = await summarizeTranscript(guildId, channelId, sessionId, sessionLogs);

    if (!summary) {
      return { success: false, summaryPath: null, error: 'Summarization returned empty' };
    }

    // Generate Word document
    const now = new Date();
    const baseFileName = `Meeting_Minutes_${now.getFullYear()}_${String(now.getMonth() + 1).padStart(2, '0')}_${String(now.getDate()).padStart(2, '0')}__${String(now.getHours()).padStart(2, '0')}_${String(now.getMinutes()).padStart(2, '0')}_recovered`;

    const wordFileName = `${baseFileName}.docx`;
    const wordPath = path.join(SUMMARY_DIR, wordFileName);

    const titleMatch = summary.match(/Thema.*?-->(.*?)<!--/);
    const meetingTitle = titleMatch ? titleMatch[1].trim() : "Meeting";

    const wordBuffer = await convertToWordDoc(summary, meetingTitle);

    if (wordBuffer) {
      fs.writeFileSync(wordPath, wordBuffer);
      markSessionSummarized(sessionId, wordPath);

      console.log(`📄 Recovery summary saved: ${wordFileName}`);

      return { success: true, summaryPath: wordPath };
    }

    return { success: false, summaryPath: null, error: 'Word document generation failed' };

  } catch (error) {
    console.error('❌ Summary generation error:', error.message);
    return { success: false, summaryPath: null, error: error.message };
  }
}

/**
 * Get recovery status
 */
export function getRecoveryStatus() {
  const sessions = getSessionsNeedingRecovery();
  let totalUntranscribed = 0;

  for (const session of sessions) {
    totalUntranscribed += getUntranscribedEntries(session.sessionId).length;
  }

  return {
    isRecovering,
    sessionsNeedingRecovery: sessions.length,
    totalUntranscribed
  };
}

/**
 * Schedule periodic recovery checks
 * @param {number} intervalMs - Check interval in milliseconds (default: 30 minutes)
 */
export function schedulePeriodicRecovery(intervalMs = 30 * 60 * 1000) {
  console.log(`📅 Scheduling periodic recovery every ${intervalMs / 60000} minutes`);

  setInterval(async () => {
    // First check for stale sessions that might have been abandoned
    markStaleSessions(2);

    const status = getRecoveryStatus();
    if (status.totalUntranscribed > 0 && !status.isRecovering) {
      console.log(`\n🔄 Periodic recovery check: ${status.totalUntranscribed} files pending`);
      await runRecovery(true);
    }
  }, intervalMs);
}

File: summarise_bot/recovery.py
from __future__ import annotations

from datetime import datetime, timezone
from pathlib import Path

from .config import SETTINGS
from .document import convert_to_word_doc, extract_meeting_title
from .logger import logger
from .manifest import (
    build_transcript_from_manifest,
    cleanup_old_sessions,
    end_session,
    get_session_entry_span,
    get_session_stats,
    get_sessions_needing_recovery,
    get_untranscribed_entries,
    is_session_fully_transcribed,
    mark_session_summarized,
    mark_stale_sessions,
    purge_transcribed_audio,
)
from .summarization import summarize_transcript
from .transcription import transcribe_with_retry
from .transcript import TranscriptStore


class RecoveryService:
    def __init__(self, transcript_store: TranscriptStore) -> None:
        self.transcript_store = transcript_store
        self.is_recovering = False

    async def run(self, auto_summarize: bool = True) -> dict[str, int]:
        if self.is_recovering:
            return {"recovered": 0, "failed": 0, "summarized": 0}
        self.is_recovering = True
        recovered = failed = summarized = 0
        try:
            cleanup_old_sessions(7)
            purge_transcribed_audio(SETTINGS.audio_retention_days)
            sessions = get_sessions_needing_recovery()
            for session in sessions:
                for entry in get_untranscribed_entries(session["sessionId"]):
                    result = await transcribe_with_retry(
                        entry["audioPath"],
                        entry["sessionId"],
                        entry["userId"],
                        entry["id"],
                        3,
                    )
                    if result.get("success"):
                        recovered += 1
                    else:
                        failed += 1
                        if result.get("isQuotaError"):
                            break
                if (
                    auto_summarize
                    and is_session_fully_transcribed(session["sessionId"])
                    and self._is_summarizable(session["sessionId"])
                ):
                    summary = await self._summarize_session(session["sessionId"])
                    if summary:
                        summarized += 1
            return {"recovered": recovered, "failed": failed, "summarized": summarized}
        finally:
            self.is_recovering = False

    def _is_summarizable(self, session_id: str) -> bool:
        """Refuse to auto-summarize a session that is not a single meeting.

        sessionId is "guild:channel", so a long-lived channel accumulates every meeting into
        one session record — 1283 entries spanning 42 days was the observed state here.
        Summarizing that produces minutes for seven meetings at once, so require the entries
        to sit inside one window and to be recent enough to still be the meeting that just ended.
        """
        oldest, newest = get_session_entry_span(session_id)
        if oldest is None or newest is None:
            return True
        span_hours = (newest - oldest).total_seconds() / 3600
        age_hours = (datetime.now(timezone.utc) - newest).total_seconds() / 3600
        max_age = SETTINGS.spool_max_age_hours
        if span_hours <= SETTINGS.session_max_span_hours and age_hours <= max_age:
            return True
        logger.warning(
            "Skipping auto-summary for a session that spans more than one meeting",
            extra={
                "action": "session_recovery",
                "event": "summary_skipped_span",
                "session_id": session_id,
                "span_hours": round(span_hours, 1),
                "age_hours": round(age_hours, 1),
                "max_span_hours": SETTINGS.session_max_span_hours,
                "max_age_hours": max_age,
            },
        )
        return False

    async def _summarize_session(self, session_id: str) -> str | None:
        transcript = build_transcript_from_manifest(session_id)
        if not transcript.strip():
            return None
        guild_id, channel_id = session_id.split(":", 1)
        ts = datetime.now(timezone.utc).replace(microsecond=0).isoformat().replace(":", "-")
        transcript_path = SETTINGS.transcript_dir / f"{guild_id}-{channel_id}-{ts}-recovered.log"
        transcript_path.write_text(transcript, encoding="utf-8")
        end_session(session_id, str(transcript_path))
        summary = await summarize_transcript(transcript_path)
        if not summary:
            return None
        file_name = f"Meeting_Minutes_{datetime.now().strftime('%Y_%m_%d__%H_%M')}_recovered.docx"
        output_path = SETTINGS.summary_dir / file_name
        title = extract_meeting_title(summary)
        word_buffer = convert_to_word_doc(summary, title)
        if word_buffer is None:
            return None
        output_path.write_bytes(word_buffer)
        mark_session_summarized(session_id, str(output_path))
        return str(output_path)

    def get_status(self) -> dict[str, int | bool]:
        sessions = get_sessions_needing_recovery()
        total_untranscribed = sum(len(get_untranscribed_entries(session["sessionId"])) for session in sessions)
        return {
            "isRecovering": self.is_recovering,
            "sessionsNeedingRecovery": len(sessions),
            "totalUntranscribed": total_untranscribed,
        }


